sql: add system table for external statement hints

DrewKimball opened the pull request:
#### sql: add hints package

This commit adds a new `sql/hints` package along with a `StatementHints`
protobuf struct and some utility functions for working with it.

Epic: None

Release note: None

#### sql: add system.statement_hints table

This patch adds a new table, `system.statement_hints`, which can be used to
associate "external" hints with a statement without having to modify
the statement itself. A following PR will add a per-node cache for this
table, similar to the way table stats are handled.

Informs #148160

Release note: None

michae2 reviewed (commented):
Very nice!

Also I didn't realize adding a system table would cause so many test changes, lol.

> @michae2 reviewed 4 of 4 files at r1, 5 of 5 files at r2, 59 of 99 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153475#-:-O_-DRws22d7_aV5E07T:bk8hr0v)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dhartunian and @kev-cao)

___
*[`pkg/sql/hints/plan_hint.go` line 36 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/153475#-O_-BKwt3xzsIphbmji7:-O_-BKwt3xzsIphbmji8:b-9tgkrr) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb93fef89e1a4e55531fd7a8f8766636c95a1133/pkg/sql/hints/plan_hint.go#L36)):*
> ```Go
> // FingerprintHashForPlanHints returns a 64-bit hash for the given query
> // fingerprint to be used in the system.plan_hints table.
> func FingerprintHashForPlanHints(queryFingerprint string) int64 {
> ```

We might want to call (or directly use) `ConstructStatementFingerprintID` here so that these fingerprint hashes match the ones used by statement_statistics.
___
*[`pkg/sql/catalog/systemschema/system.go` line 1393 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/153475#-O_-CEUS2T1X1j8K4Ma-:-O_-CEUS2T1X1j8K4Ma0:bawlor9) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c9a478a61f8af834db5e9b4cb781dbb157e3c45/pkg/sql/catalog/systemschema/system.go#L1393)):*
> ```Go
> 	PlanHintsTableSchema = `
>   CREATE TABLE system.plan_hints (
>     "query_hash"  INT8 NOT VISIBLE NOT NULL AS (crdb_internal.gen_query_fingerprint_hash(fingerprint)) STORED,
> ```

We might consider calling this `fingerprint_id` to match system.statement_statistics (if it is the same hash).
___
*[`pkg/sql/catalog/systemschema/system.go` line 1398 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/153475#-O_-ClcY-wtNyso_K1M_:-O_-ClcY-wtNyso_K1Ma:b-97h6p6) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c9a478a61f8af834db5e9b4cb781dbb157e3c45/pkg/sql/catalog/systemschema/system.go#L1398)):*
> ```Go
>     "plan_hint"   BYTES NOT NULL,
>     "created_at"  TIMESTAMPTZ NOT NULL DEFAULT now(),
>     CONSTRAINT "primary" PRIMARY KEY ("query_hash", "row_id"),
> ```

Do we need query_hash to be part of the PK?

Maybe a secondary index on query_hash instead?


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
>Also I didn't realize adding a system table would cause so many test changes, lol.

Yeah kind of crazy...

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153475#-:-O_0LL30-LHByofrYtOh:b34l9jy)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dhartunian, @kev-cao, and @michae2)

___
*[`pkg/sql/catalog/systemschema/system.go` line 1398 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/153475#-O_-ClcY-wtNyso_K1M_:-O_0Iuzp7QxyRZn1_s_b:b-3cqqan) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c9a478a61f8af834db5e9b4cb781dbb157e3c45/pkg/sql/catalog/systemschema/system.go#L1398)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Do we need query_hash to be part of the PK?

Maybe a secondary index on query_hash instead?
</blockquote></details>

It doesn't have to be, but it doesn't seem really useful to have a PK organized by row ID when we'll be querying by fingerprint/hash. Is the thinking that we might want to be able to handle rows by a single int instead of two?


<!-- Sent from Reviewable.io -->

DrewKimball committed:
sql: add hints package

This commit adds a new `sql/hints` package along with a `StatementHints`
protobuf struct and some utility functions for working with it.

Epic: None

Release note: None

DrewKimball committed:
sql: add system.statement_hints table

This patch adds a new table, `system.statement_hints`, which can be used to
associate "external" hints with a statement without having to modify
the statement itself. A following PR will add a per-node cache for this
table, similar to the way table stats are handled.

Informs #148160

Release note: None

DrewKimball reviewed (commented):
I've updated the table to be called `statement_hints` and changed all references to "plan hints" to be "statement hints" instead, reflecting the fact that these hints may have a broader effect than just modifying or fixing a plan. I've also updated the table schema so that there is a row ID primary key, and a secondary index on the computed hash column.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153475#-:-O_Ftg79EKA1wFVCBchX:bfr82jv)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dhartunian, @kev-cao, and @michae2)

___
*[`pkg/sql/catalog/systemschema/system.go` line 1393 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/153475#-O_-CEUS2T1X1j8K4Ma-:-O_FtYeK2B06-TgXu-K-:b-a3jwrh) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c9a478a61f8af834db5e9b4cb781dbb157e3c45/pkg/sql/catalog/systemschema/system.go#L1393)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

We might consider calling this `fingerprint_id` to match system.statement_statistics (if it is the same hash).
</blockquote></details>

As we discussed offline, this will be a different hash than the one for statement statistics.
___
*[`pkg/sql/hints/plan_hint.go` line 36 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/153475#-O_-BKwt3xzsIphbmji7:-O_Ft6LR3tz6p-CQvPcf:b-uc7eal) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb93fef89e1a4e55531fd7a8f8766636c95a1133/pkg/sql/hints/plan_hint.go#L36)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

We might want to call (or directly use) `ConstructStatementFingerprintID` here so that these fingerprint hashes match the ones used by statement_statistics.
</blockquote></details>

I've realized we can just call the existing `fnv64` builtin, so we do that now. Also, we want the bucketing for this table to be broader than that used for statement statistics, which includes the DB and whether the transaction is explicit. (We discussed that offline but repeating here for posterity).


<!-- Sent from Reviewable.io -->

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice!

> @michae2 reviewed 108 of 108 files at r4, 109 of 109 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153475#-:-O_QURZHACfOjteurzVf:b-8gk8uj)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @dhartunian and @kev-cao)

___
*[`pkg/upgrade/upgrades/v25_4_system_statement_hints.go` line 27 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/153475#-O_QZWM_74BFg9AV8ZLW:-O_QZWM_74BFg9AV8ZLX:b-gg129u) ([raw file](https://github.com/cockroachdb/cockroach/blob/e389d75d3cdb6c945925acb242c1c3594fffc0ec/pkg/upgrade/upgrades/v25_4_system_statement_hints.go#L27)):*
> ```Go
> ) error {
> 	return createSystemTable(
> 		ctx, d.DB, d.Settings, d.Codec, systemschema.StatementHintsTable, tree.LocalityLevelTable,
> ```

Should this be LocalityLevelGlobal?


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
TFYR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153475#-:-O_TkqqR9kzSLeO30Bhh:b-814cj3)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @dhartunian, @kev-cao, and @michae2)

___
*[`pkg/upgrade/upgrades/v25_4_system_statement_hints.go` line 27 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/153475#-O_QZWM_74BFg9AV8ZLW:-O_TkNyf8osdI_EqiwxJ:bgfciar) ([raw file](https://github.com/cockroachdb/cockroach/blob/e389d75d3cdb6c945925acb242c1c3594fffc0ec/pkg/upgrade/upgrades/v25_4_system_statement_hints.go#L27)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Should this be LocalityLevelGlobal?
</blockquote></details>

I considered that, but I don't think there's currently a precedent for a system table being global by default, and there was already some work done to make it possible to change the locality for https://github.com/cockroachdb/cockroach/issues/63365. I think it will be better for this table to piggy-back off of that work, rather than do something new. What do you think?


<!-- Sent from Reviewable.io -->

michae2 reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153475)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @dhartunian, @DrewKimball, and @kev-cao)

___
*[`pkg/upgrade/upgrades/v25_4_system_statement_hints.go` line 27 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/153475#-O_QZWM_74BFg9AV8ZLW:-O_VCpU0DrBMMRtHtFLo:bpm9lsw) ([raw file](https://github.com/cockroachdb/cockroach/blob/e389d75d3cdb6c945925acb242c1c3594fffc0ec/pkg/upgrade/upgrades/v25_4_system_statement_hints.go#L27)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

I considered that, but I don't think there's currently a precedent for a system table being global by default, and there was already some work done to make it possible to change the locality for https://github.com/cockroachdb/cockroach/issues/63365. I think it will be better for this table to piggy-back off of that work, rather than do something new. What do you think?
</blockquote></details>

Sounds fine to me.


<!-- Sent from Reviewable.io -->

DrewKimball commented:
bors r=michae2
